### PR TITLE
[CALCITE-6812] Support base64 and unbase64 functions for Hive

### PR DESCRIPTION
**Pull request by xuzifu666**
Currently calcite not support base64 and unbase64 functions for Hive Library, need to support it.
jira: https://issues.apache.org/jira/browse/CALCITE-6812

### CODE REVIEW

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 1]
```diff
@@ -5173,6 +5173,26 @@ void testBitGetFunc(SqlOperatorFixture f, String functionName) {
     f.checkString("to_base64(x'61')", "YQ==", "VARCHAR NOT NULL");
   }
 
+  @Test void testBase64() {
```
Comment (caicancai) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:5176]: Add jira link

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 2]
```diff
@@ -5173,6 +5173,26 @@ void testBitGetFunc(SqlOperatorFixture f, String functionName) {
...
+    final SqlOperatorFixture f = fixture().withLibrary(SqlLibrary.HIVE);
+    f.setFor(SqlLibraryOperators.BASE64);
+    f.checkString("base64(cast('a' as binary))", "YQ==", "VARCHAR NOT NULL");
+    f.checkString("base64('')", "", "VARCHAR NOT NULL");
```
Comment (caicancai) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:5180]: how about base64(null)

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 1 above)
Comment (xuzifu666) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:5179]: Done, Thanks @caicancai

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 2 above)
Comment (xuzifu666) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:5183]: It would be null and had added case to it.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 3]
```diff
@@ -330,6 +330,28 @@ private static String toBase64_(byte[] bytes) {
     }
   }
 
+  /** SQL BASE64(string) function for string. */
+  public static String base64(String string) {
```
Comment (ILuffZhe) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:334]: There already exists `toBase64(ByteString string)` and `toBase64(String string)` in SqlFunctions.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 4]
```diff
@@ -330,6 +330,28 @@ private static String toBase64_(byte[] bytes) {
...
+  }
+
+  /** SQL BASE64(string) function for bytestring. */
+  public static String base64(ByteString byteString) {
+    byte[] bytes = byteString.getBytes();
+    return toBase64_(bytes);
+  }
+
+  /** SQL UNBASE64(string) function. */
+  public static @Nullable String unBase64(String base64) {
```
Comment (ILuffZhe) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:345]: Can we make it return String, so `ByteString fromBase64(String base64)` can be used.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 5]
```diff
@@ -330,6 +330,28 @@ private static String toBase64_(byte[] bytes) {
...
+    try {
+      base64 = FROM_BASE64_REGEXP.matcher(base64).replaceAll("");
+      //CHECKSTYLE: IGNORE 1
```
Comment (ILuffZhe) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:348]: What's this comment for?

Comment (xuzifu666) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:348]: Calcite can not support new String method，use this comment to suppress it. Other place in Calcite also use it.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 6]
```diff
@@ -5173,6 +5173,31 @@ void testBitGetFunc(SqlOperatorFixture f, String functionName) {
...
+
+  @Test void testUnBase64() {
+    final SqlOperatorFixture f = fixture().withLibrary(SqlLibrary.HIVE);
+    f.setFor(SqlLibraryOperators.UN_BASE64);
+    f.checkString("unbase64('VGhpcyBpcyBhIHRlc3QgU3RyaW5nLg==')",
+              "This is a test String.",
+              "VARCHAR");
+    f.checkString("unbase64('YQ==')", "a", "VARCHAR");
+    f.checkString("unbase64('')", "", "VARCHAR");
+    f.checkNull("unbase64(null)");
```
Comment (ILuffZhe) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:5198]: Can you add a test to check `IllegalArgumentException` with null result for unbase64.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 7]
```diff
@@ -1855,6 +1855,19 @@ private static RelDataType deriveTypeMapFromEntries(SqlOperatorBinding opBinding
           OperandTypes.STRING.or(OperandTypes.BINARY),
           SqlFunctionCategory.STRING);
 
+  @LibraryOperator(libraries = {HIVE})
+  public static final SqlFunction BASE64 =
+      SqlBasicFunction.create("BASE64",
+          ReturnTypes.VARCHAR_NULLABLE,
```
Comment (caicancai) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1861]: Why is one return value VARCHAR_NULLABLE and the other VARCHAR_FORCE_NULLABLE? Is there any difference between base64 and unbase64?

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 6 above)
Comment (xuzifu666) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:5198]: OK, would do it.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 3 above)
Comment (xuzifu666) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:334]: You mean here reuse toBase64 method?  It is OK for it.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 7 above)
Comment (xuzifu666) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1861]: OK, I would test it.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 3 above)
Comment (ILuffZhe) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:334]: Yes, it would be much better if we can reuse or add arguments to existing methods.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 4 above)
Comment (xuzifu666) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Done

**Comment by xuzifu666** [comment]
@caicancai @ILuffZhe had addressed comments, PTAL when had free time, thanks~

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 3 above)
Comment (xuzifu666) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Done

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 7 above)
Comment (xuzifu666) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1861]: According to @ILuffZhe suggestions: BASE64 reuse TO_BASE64 function and UN_BASE64 reuse FROM_BASE64, so ReturnTypes keep the same with TO_BASE64 and UN_BASE64 , they are VARCHAR_NULLABLE and VARBINARY_FORCE_NULLABLE which had the same behavior in Hive. @caicancai

**Review by caicancai — APPROVED** [review]
LGTM

**Review by ILuffZhe — APPROVED** [review]

**Comment by xuzifu666** [comment]
Very thanks for @caicancai @ILuffZhe review~
